src/planning.py

```diff
@@ -272,29 +295,33 @@ def _run_low_level_search(task: Task, option_model: _OptionModelBase,
                 discovered_failures[cur_idx] = failure
                 # If we're immediately propagating failures, raise it now.
                 if CFG.sesame_propagate_failures == "immediately":
-                    raise _DiscoveredFailureException("Discovered a failure",
-                                                      failure)
+                    raise _DiscoveredFailureException(
+                        "Discovered a failure", failure,
+                        {"longest_refinement": longest_refinement})
             if not discovered_failures[cur_idx]:
                 traj[cur_idx + 1] = next_state
                 cur_idx += 1
                 # Check atoms against expected atoms_sequence constraint.
                 assert len(traj) == len(atoms_sequence)
                 # The expected atoms are ones that we definitely expect to be
-                # true at this point in the plan. They are not *all* the atoms
-                # that could be true.
+                # true at this point in the refinement. They are not *all* the
+                # atoms that could be true.
                 expected_atoms = {
                     atom
                     for atom in atoms_sequence[cur_idx]
                     if atom.predicate.name != _NOT_CAUSES_FAILURE
                 }
-                # This is equivalent to, but faster than, checking whether
-                # expected_atoms is a subset of utils.abstract(traj[cur_idx],
-                # predicates).
+                # This "if all" statement is equivalent to, but faster than,
+                # checking whether expected_atoms is a subset of
+                # utils.abstract(traj[cur_idx], predicates).
                 if all(atom.holds(traj[cur_idx]) for atom in expected_atoms):
                     can_continue_on = True
-                    if cur_idx == len(skeleton):  # success!
-                        result = plan
-                        return result
+                    # Update the longest_refinement found so far.
```

Comment:
This logic was up higher earlier, which I think was wrong. We only want to update the longest_refinement after you pass the expected atoms check at any step.

Reply:
i'm not sure if this is what we want. i think it is helpful to see where the refinement is getting stuck. for example, in the option learning stuff, it is informative to see the option trying to do something but failing (e.g. trying to pick or trying to place). we can rename or make clearer somehow, but I do think we want the original implementation.